source/adapters/level_zero/command_buffer.cpp

```diff
@@ -437,6 +437,21 @@ urCommandBufferReleaseExp(ur_exp_command_buffer_handle_t CommandBuffer) {
 
 UR_APIEXPORT ur_result_t UR_APICALL
 urCommandBufferFinalizeExp(ur_exp_command_buffer_handle_t CommandBuffer) {
+  // Reset the L0 events we use for command-buffer internal sync-points to the
+  // non-signalled state
+  ZE2UR_CALL(zeCommandListAppendBarrier,
+             (CommandBuffer->ZeCommandList, nullptr, 0, nullptr));
+  for (auto SyncPoint : CommandBuffer->SyncPoints) {
+    ur_event_handle_t Event = SyncPoint.second;
+    ZE2UR_CALL(zeCommandListAppendEventReset,
+               (CommandBuffer->ZeCommandList, Event->ZeEvent));
+  }
+
+  // Reset the wait-event for the UR command-buffer that is signalled when its
+  // submission dependencies have been satisfied.
+  ZE2UR_CALL(zeCommandListAppendEventReset,
+             (CommandBuffer->ZeCommandList, CommandBuffer->WaitEvent->ZeEvent));
+
```

Comment:
```suggestion
  size_t NumEvents = CommandBuffer->SyncPoints.size() + 1;
  std::vector<ze_event_handle_t> TmpEventList{NumEvents};

  // Reset the wait-event for the UR command-buffer that is signalled when its
  // submission dependencies have been satisfied.
  TmpEventList[0] = CommandBuffer->WaitEvent->ZeEvent;

  // Reset the L0 events we use for command-buffer internal sync-points
  for (size_t i = 1; i < NumEvents; i++) {
    TmpEventList[i] = CommandBuffer->SyncPoints[i - 1]->ZeEvent;
  }

  ZE2UR_CALL(zeCommandListAppendBarrier, (CommandBuffer->ZeCommandList, nullptr,
                                          NumEvents, TmpEventList.data()));

  for (size_t i = 0; i < NumEvents; i++) {
    ZE2UR_CALL(zeCommandListAppendEventReset,
               (CommandBuffer->ZeCommandList, TmpEventList[i]));
  }
```

Reply:
Thanks for the patch, I used it for inspiration but didn't apply it directly, as I don't think it handles ` CommandBuffer->WaitEvent` correctly. We already have a barrier command to wait on this event at the start of the command-list (added in `urCommandBufferCreateExp`), so don't need to wait on it again. We also need to reset this, which is code that's removed by this patch.